File: algorithms/go/278-first bad version/first-bad-version.go
package problem278

// binary search
func firstBadVersion(n int) int {
	lo, hi := 0, n
	for lo < hi {
		var mid = (hi-lo)/2 + lo
		if isBadVersion(mid) {
			hi = mid
		} else {
			lo = mid + 1
		}
	}
	return hi
}

func isBadVersion(version int) bool {
	return true
}

File: algorithms/278-first-bad-version/first-bad-version.py
"""
278. first bad version

https://leetcode.com/problems/first-bad-version/
https://leetcode-cn.com/problems/first-bad-version/

第一个错误的版本
"""


# The isBadVersion API is already defined for you.
# @param version, an integer
# @return an integer
# def isBadVersion(version):

class Solution:
    def first_bad_version(self, n):
        """
        binary search
        """
        lo, hi = 1, n
        while lo < hi:
            mid = (hi - lo) // 2 + lo
            if isBadVersion(mid):
                hi = mid
            else:
                lo = mid + 1

        return hi
